Client/Frontend/Library/HistoryPanel/HistoryPanel.swift

```diff
@@ -0,0 +1,651 @@
+    }
+    
+    // MARK: - Swipe Action helpers
+    
+    func tableView(_ tableView: UITableView, trailingSwipeActionsConfigurationForRowAt indexPath: IndexPath) -> UISwipeActionsConfiguration? {
+        let cell = diffableDatasource?.itemIdentifier(for: indexPath)
+        
+        if let _ = cell as? Site {
+            let deleteAction = UIContextualAction(style: .destructive, title: .HistoryPanelDelete) { [weak self] (_, _, completion) in
+                guard let self = self else {
+                    // OSLOG - couldn't weakify self!
+                    completion(false)
+                    return
+                }
+
+                self.removeHistoryItem(at: indexPath)
+                completion(true)
+            }
+
+            return UISwipeActionsConfiguration(actions: [deleteAction])
+        }
+        
+        return nil
+    }
+
+    // MARK: - Empty State helpers
+    
+    private func updateEmptyPanelState() {
+        if viewModel.groupedSites.isEmpty {
+            if emptyStateOverlayView.superview == nil {
+                tableView.tableFooterView = emptyStateOverlayView
+            }
+        } else {
+            tableView.alwaysBounceVertical = true
+            tableView.tableFooterView = nil
+        }
+    }
+
+    private func createEmptyStateOverlayView() -> UIView {
+        let overlayView = UIView()
+
+        // overlayView becomes the footer view, and for unknown reason, setting the bgcolor is ignored.
+        // Create an explicit view for setting the color.
+        let bgColor = UIView()
+        bgColor.backgroundColor = UIColor.theme.homePanel.panelBackground
+        overlayView.addSubview(bgColor)
+        bgColor.snp.makeConstraints { make in
+            // Height behaves oddly: equalToSuperview fails in this case, as does setting top.equalToSuperview(), simply setting this to ample height works.
+            make.height.equalTo(UIScreen.main.bounds.height)
+            make.width.equalToSuperview()
+        }
+
+        let welcomeLabel = UILabel()
+        overlayView.addSubview(welcomeLabel)
+        welcomeLabel.text = .HistoryPanelEmptyStateTitle
+        welcomeLabel.textAlignment = .center
+        welcomeLabel.font = DynamicFontHelper.defaultHelper.DeviceFontLight
+        welcomeLabel.textColor = UIColor.theme.homePanel.welcomeScreenText
+        welcomeLabel.numberOfLines = 0
+        welcomeLabel.adjustsFontSizeToFitWidth = true
+
+        welcomeLabel.snp.makeConstraints { make in
+            make.centerX.equalTo(overlayView)
+            // Sets proper top constraint for iPhone 6 in portait and for iPad.
+            make.centerY.equalTo(overlayView).offset(LibraryPanelUX.EmptyTabContentOffset).priority(100)
+            // Sets proper top constraint for iPhone 4, 5 in portrait.
+            make.top.greaterThanOrEqualTo(overlayView).offset(50)
+            make.width.equalTo(HistoryPanelUX.WelcomeScreenItemWidth)
+        }
```

Comment:
snapkit?

Reply:
I just moved code around here. Will pick away at these things in a lighter sprint!